activity_calendar/api.py

```diff
@@ -69,3 +73,53 @@ def fullcalendar_feed(request):
             activity_moment_jsons.append(json_instance)
 
     return JsonResponse({'activities': activity_moment_jsons})
+
+
+@require_safe
+def upcoming_core_feed(request):
+    """
+        A collection of the first occurrence of all core activities. E.g. the first
+        upcoming boardgame-related activity, the first upcoming swordfighting training, etc.
+        Skips cancelled and removed activities.
+        Used on kotkt.nl
+    """
+    grouping_identifiers = request.GET.get('groups', "").split(',')
+
+    now = timezone.now()
+
+    # Add placeholders to ensure the order of the groupings does not change
+    earliest_moments = {
+        x:None for x in grouping_identifiers
+    }
+
+    # Iterate over all published activities that are part of the core groupings in the GET request
+    for activity in Activity.objects.select_related('core_grouping').filter(published_date__lte=now, core_grouping__identifier__in=grouping_identifiers):
+        # Fetch the next activitymoment (if any) for this activity that is not cancelled
+        activity_moment = activity.get_next_activitymoment(dtstart=now, exclude_cancelled=True)
+        earliest_moment = earliest_moments.get(activity.core_grouping.identifier, None)
+        # Does this activity take place earlier than the current earliest activitymoment for this grouping?
+        if earliest_moment is None \
+                or earliest_moment.start_date > activity_moment.start_date:
+            earliest_moments[activity.core_grouping.identifier] = activity_moment
+
+    # Fetch data for the selected activitymoments
+    activity_moment_jsons = []
+    for activity_moment in earliest_moments.values():
+        if activity_moment is None:
+            continue
+
```

Comment:
Hmm, I think I accidentally lost this remark when submitting the review  again. So I'll type it again:
Whereas earlier you try to maintain order of the grouping identifiers, the order is completely rejected at this point. Particularly when an activitymoment was not found it just skips a step in the list. The json data also does not communicate the relevant grouping  identifiers. Could you rewrite the code to either just receive one identifier or, when allowing multiple to be added, have the result contain clarity on which activitymoment relates to which identifier.

Reply:
The skipping was done on purpose (as activities might have been returned for other groupings in the same request). Added the grouping identifier in the returned activitymoment data 👍